[Merged by Bors] - feat(linear_algebra/lagrange): Refactor lagrange interpolation and add extended proofs.

linesthatinterlace opened the pull request:
The current formulation of Lagrange interpolants phrases things in terms of a finset of nodes, and functions interpolated at those nodes. This changes things in order to allow for indexed nodes and weights. It also extends various functionalities of the lagrange interpolation.

---

It is more standard - and in many ways more natural - to think in terms of nodes and weights (that is, evaluations) at those nodes. This corresponds to a reformulation in terms of injective functions from fintypes. This commit implements the necessary changes to do this, and attempts to introduce API to fill the gaps left behind.

I am looking for feedback on this in terms of overall approach as well as details - I should point out that this isn't for the sake of it: in the real context of the theory of binary Goppa codes, I was encountering issues with the previous formulation as it is very natural to want to index your weights by the same thing that one indexes your nodes: and, again, it's natural for these to be tuples.

This new approach seems to give tighter and neater proofs, which seems a good sign.

linesthatinterlace committed:
Add useful lemmas linking the Vandermonde to polynomials

linesthatinterlace committed:
Refactor the lagrange interpolation in terms of fintypes

vihdzp commented on src/linear_algebra/lagrange.lean:
```diff
@@ -1,224 +1,353 @@
 end
 
-@[simp] theorem eval_basis_ne (x y : F) (h1 : y ∈ s) (h2 : y ≠ x) : (basis s x).eval y = 0 :=
+lemma node_linear_zero_iff : node_linear v i j = 0 ↔ i = j :=
```
```suggestion
@[simp] lemma node_linear_zero_iff : node_linear v i j = 0 ↔ i = j :=
```

vihdzp commented on src/linear_algebra/lagrange.lean:
```diff
@@ -1,224 +1,353 @@
 
-theorem eval_basis (x y : F) (h : y ∈ s) : (basis s x).eval y = if y = x then 1 else 0 :=
-by { split_ifs with H, { subst H, apply eval_basis_self }, { exact eval_basis_ne s x y h H } }
+lemma node_linear_degree_eq : (node_linear v i i).degree = ⊥ :=
```
```suggestion
@[simp] lemma node_linear_degree_eq : (node_linear v i i).degree = ⊥ :=
```

vihdzp commented on src/linear_algebra/lagrange.lean:
```diff
@@ -1,224 +1,353 @@
+lemma node_linear_nat_degree : (node_linear v i j).nat_degree = ite (i = j) 0 1 :=
+by { rw [nat_degree, node_linear_degree], split_ifs; refl }
+
+lemma eval_node_linear_right : eval (v j) (node_linear v i j) = 0 :=
```
```suggestion
@[simp] lemma eval_node_linear_right : eval (v j) (node_linear v i j) = 0 :=
```

vihdzp commented on src/linear_algebra/lagrange.lean:
```diff
@@ -1,224 +1,353 @@
+
+theorem basis_eq_of_is_empty [is_empty ι] : basis v i = 1 := basis_eq_of_subsingleton
+
+lemma basis_ne_zero : basis v i ≠ 0 :=
```
```suggestion
@[simp] lemma basis_ne_zero : basis v i ≠ 0 :=
```

vihdzp commented on src/linear_algebra/lagrange.lean:
```diff
@@ -1,224 +1,353 @@
+theorem basis_eq_of_subsingleton [subsingleton ι] : basis v i = 1 :=
+by simp_rw [basis_def, eq_iff_true_of_subsingleton, if_true, prod_const_one]
+
+theorem basis_eq_of_unique [unique ι] : basis v i = 1 := basis_eq_of_subsingleton
```
```suggestion
@[simp] theorem basis_eq_of_unique [unique ι] : basis v i = 1 := basis_eq_of_subsingleton
```

vihdzp commented on src/linear_algebra/lagrange.lean:
```diff
@@ -1,224 +1,353 @@
+
+theorem basis_eq_of_unique [unique ι] : basis v i = 1 := basis_eq_of_subsingleton
+
+theorem basis_eq_of_is_empty [is_empty ι] : basis v i = 1 := basis_eq_of_subsingleton
```
```suggestion
@[simp] theorem basis_eq_of_is_empty [is_empty ι] : basis v i = 1 := basis_eq_of_subsingleton
```

vihdzp commented on src/linear_algebra/lagrange.lean:
```diff
@@ -1,224 +1,353 @@
+theorem eval_basis : (basis v i).eval (v j) = if i = j then 1 else 0 :=
+by { split_ifs with H, { subst H, exact eval_basis_self}, { exact eval_basis_ne H } }
+
+theorem nat_degree_basis : (basis v i).nat_degree = (fintype.card ι) - 1 :=
```
```suggestion
@[simp] theorem nat_degree_basis : (basis v i).nat_degree = (fintype.card ι) - 1 :=
```

vihdzp commented on src/linear_algebra/lagrange.lean:
```diff
@@ -1,224 +1,353 @@
-eq_of_eval_eq s (degree_interpolate_lt _ _) hg $ λ x hx, begin
-  rw hgf x hx,
-  exact eval_interpolate _ _ _ hx,
+theorem degree_basis : (basis v i).degree = ↑(fintype.card ι - 1) :=
```
```suggestion
@[simp] theorem degree_basis : (basis v i).degree = ↑(fintype.card ι - 1) :=
```

vihdzp commented on src/linear_algebra/lagrange.lean:
```diff
@@ -1,224 +1,353 @@
+
+lemma interpolate_def : interpolate v r = ∑ i, C (r i) * basis v i := rfl
+
+theorem interpolate_empty [is_empty ι] : interpolate v r = 0 :=
```
```suggestion
@[simp] theorem interpolate_empty [is_empty ι] : interpolate v r = 0 :=
```

vihdzp commented on src/linear_algebra/lagrange.lean:
```diff
@@ -1,224 +1,353 @@
+theorem interpolate_singleton [unique ι] : interpolate v r = C (r default) :=
+by rw [interpolate_def, univ_unique, sum_singleton, basis_eq_of_unique, mul_one]
+
+theorem eval_interpolate : eval (v i) (interpolate v r) = r i :=
```
```suggestion
@[simp] theorem eval_interpolate : eval (v i) (interpolate v r) = r i :=
```

vihdzp commented on src/linear_algebra/lagrange.lean:
```diff
@@ -1,224 +1,353 @@
+lemma interpolate_add (r s) :
+ interpolate v (r + s) = interpolate v r + interpolate v s := (linterpolate v).map_add r s
+
+lemma interpolate_zero : interpolate v 0 = 0 :=
```
```suggestion
@[simp] lemma interpolate_zero : interpolate v 0 = 0 :=
```

vihdzp commented on src/linear_algebra/lagrange.lean:
```diff
@@ -1,224 +1,353 @@
+lemma interpolate_zero : interpolate v 0 = 0 :=
+(linterpolate v).map_zero
+
+lemma interpolate_neg (r) : interpolate v (-r) = -interpolate v r :=
```
```suggestion
@[simp] lemma interpolate_neg (r) : interpolate v (-r) = -interpolate v r :=
```

vihdzp commented on src/linear_algebra/lagrange.lean:
```diff
@@ -1,224 +1,353 @@
+lemma interpolate_neg (r) : interpolate v (-r) = -interpolate v r :=
+(linterpolate v).map_neg r
+
+lemma interpolate_sub (r s) :
```
```suggestion
@[simp] lemma interpolate_sub (r s) :
```

vihdzp commented on src/linear_algebra/lagrange.lean:
```diff
@@ -1,224 +1,353 @@
+lemma interpolate_sub (r s) :
+ interpolate v (r - s) = interpolate v r - interpolate v s := (linterpolate v).map_sub r s
+
+lemma interpolate_smul (c : F) (r) :
```
```suggestion
@[simp] lemma interpolate_smul (c : F) (r) :
```

vihdzp commented on src/linear_algebra/lagrange.lean:
```diff
@@ -1,224 +1,353 @@
+
+lemma interpolate_at_def : interpolate_at v f = interpolate v (λ i, f (v i)) := rfl
+
+theorem interpolate_at_eq_of_eval_eq (hs : ∀ i, f (v i) = g (v i)) :
```
Maybe this could be an `ext` lemma?

wrenna-robson replied:
I considered this - clearly it's doing that kind of thing. `interpolate_eq_of_eval_eq` should probably also be. Is this the standard form?

vihdzp commented on src/linear_algebra/lagrange.lean:
```diff
@@ -1,224 +1,353 @@
+ interpolate_at v f = interpolate_at v g :=
+by {simp_rw [interpolate_at_def], congr, exact funext hs}
+
+theorem interpolate_at_empty [is_empty ι] : interpolate_at v f = 0 :=
```
```suggestion
@[simp] theorem interpolate_at_empty [is_empty ι] : interpolate_at v f = 0 :=
```

vihdzp commented on src/linear_algebra/lagrange.lean:
```diff
@@ -1,224 +1,353 @@
+theorem interpolate_at_singleton [unique ι] : interpolate_at v f = C (f (v default)) :=
+by rw [interpolate_at_def, interpolate_singleton]
+
+theorem eval_interpolate_at : eval (v i) (interpolate_at v f) = f (v i) :=
```
```suggestion
@[simp] theorem eval_interpolate_at : eval (v i) (interpolate_at v f) = f (v i) :=
```

vihdzp commented on src/linear_algebra/lagrange.lean:
```diff
@@ -1,224 +1,353 @@
+  interpolate_at v (f + g) = interpolate_at v f + interpolate_at v g :=
+(linterpolate_at v).map_add f g
+
+lemma interpolate_at_zero : interpolate_at v 0 = 0 :=
```
```suggestion
@[simp] lemma interpolate_at_zero : interpolate_at v 0 = 0 :=
```

vihdzp commented on src/linear_algebra/lagrange.lean:
```diff
@@ -1,224 +1,353 @@
+ map_add' := λ f g, by { simp_rw [interpolate_at_def, ← interpolate_add], refl },
+ map_smul' := λ c f, by { { simp_rw [interpolate_at_def, ← interpolate_smul], refl } } }
+
+lemma interpolate_at_add (f g) :
```
```suggestion
@[simp] lemma interpolate_at_add (f g) :
```

vihdzp commented on src/linear_algebra/lagrange.lean:
```diff
@@ -1,224 +1,353 @@
+lemma interpolate_at_zero : interpolate_at v 0 = 0 :=
+(linterpolate_at v).map_zero
+
+lemma interpolate_at_neg (f) : interpolate_at v (-f) = -interpolate_at v f :=
```
```suggestion
@[simp] lemma interpolate_at_neg (f) : interpolate_at v (-f) = -interpolate_at v f :=
```

vihdzp commented on src/linear_algebra/lagrange.lean:
```diff
@@ -1,224 +1,353 @@
+lemma interpolate_at_neg (f) : interpolate_at v (-f) = -interpolate_at v f :=
+(linterpolate_at v).map_neg f
+
+lemma interpolate_at_sub (f g) :
```
```suggestion
@[simp] lemma interpolate_at_sub (f g) :
```

vihdzp commented on src/linear_algebra/lagrange.lean:
```diff
@@ -1,224 +1,353 @@
+interpolate_at v (f - g) = interpolate_at v f - interpolate_at v g :=
+(linterpolate_at v).map_sub f g
+
+lemma interpolate_at_smul (c : F) (f) : interpolate_at v (c • f) = c • interpolate_at v f :=
```
```suggestion
@[simp] lemma interpolate_at_smul (c : F) (f) : interpolate_at v (c • f) = c • interpolate_at v f :=
```

vihdzp commented on src/linear_algebra/vandermonde.lean:
```diff
@@ -133,4 +134,63 @@ lemma det_vandermonde_ne_zero_iff [is_domain R] {n : ℕ} {v : fin n → R} :
+theorem vandermonde_invertibility {R : Type*} [comm_ring R]
+[is_domain R] {n : ℕ} (v : fin n ↪ R) {f : fin n → R}
+(h₂ : ∀ j, ∑ i, f i * (v j ^ (i : ℕ)) = 0) : f = 0
+:= by {refine vandermonde_invertibility' v _, simp_rw mul_comm, exact h₂}
```
```suggestion
:= by { refine vandermonde_invertibility' v _, simp_rw mul_comm, exact h₂ }
```

vihdzp commented on src/linear_algebra/vandermonde.lean:
```diff
@@ -133,4 +134,63 @@ lemma det_vandermonde_ne_zero_iff [is_domain R] {n : ℕ} {v : fin n → R} :
+:= eq_zero_of_mul_vec_eq_zero (det_vandermonde_ne_zero_iff.mpr v.inj') (funext h₂)
+
+theorem vandermonde_invertibility {R : Type*} [comm_ring R]
+[is_domain R] {n : ℕ} (v : fin n ↪ R) {f : fin n → R}
```
Why is `v` explicit but `f` implicit? I feel like due to `h₂`, it should either be both implicit or both explicit.

wrenna-robson replied:
It turns out that when you're using this in practice, often the f is clear from context but it needs help finding the v. I'd be happy to make both explicit if you think that's more to style.

vihdzp replied:
It's not a big deal, just a thing to keep in mind.

vihdzp commented on src/linear_algebra/vandermonde.lean:
```diff
@@ -133,4 +134,63 @@ lemma det_vandermonde_ne_zero_iff [is_domain R] {n : ℕ} {v : fin n → R} :
+open_locale polynomial
+
+theorem eq_zero_iff_degree_lt_equiv_eq_zero {R : Type*} [comm_ring R] {n : ℕ}
+  {p : R[X]} (hp : p ∈ degree_lt R n) : p = 0 ↔ degree_lt_equiv _ _ ⟨p, hp⟩ = 0 :=
```
The other direction might make a good `simp` lemma.

wrenna-robson replied:
Don't disagree. I actually had a bunch more of these lemmas and defined a notation for `degree_lt_equiv _ _` but it seemed overkill in the end.

vihdzp reviewed (commented):
There's tons of untapped `simp` potential.

linesthatinterlace committed:
Add missing simp lemmas

I stripped out all the simps from the original file when refactoring - this adds a chunk of them back in (though they have not been linted).

Co-authored-by: Violeta Hernández <vi.hdz.p@gmail.com>

linesthatinterlace committed:
Updated documentation, naming and whitespace

eric-wieser commented on src/linear_algebra/lagrange.lean:
```diff
@@ -1,224 +1,371 @@
+/-- The basis divisor is defined in terms of an embedded `v : ι ↪ F` from a fintype `ι`. `basis_divisor v i j` is the unique polynomial with `degree ≤ 1` such that when evaluated at `v i` it gives `1` and `v j` it gives `0` (where when `i = j` it is identically `0`).
+
+Conceptually, they are therefore the building blocks for the Lagrange interpolants. -/
+def basis_divisor (v : ι ↪ F) : ι → ι → F[X] := λ i j, C (v i - v j)⁻¹ * (X - C (v j))
```
```suggestion
Conceptually, they are therefore the building blocks for the Lagrange interpolants. -/
def basis_divisor (v : ι ↪ F) (i j : ι) : F[X] := C (v i - v j)⁻¹ * (X - C (v j))
```
then you can remove `basis_divisor_def` below as it's just `basis_divisor` (the generated lemma)

wrenna-robson replied:
I wasn't sure about when we do this kind of thing, having a def lemma that unfolds a definition. But if this'll work it's clearly better.

eric-wieser commented on src/linear_algebra/lagrange.lean:
```diff
@@ -1,224 +1,371 @@
+
+/-- Lagrange basis polynomials indexed by `ι` defined for an embedding `v : ι ↪ F`.
+`basis v i` evaluates to 1 at `v i` and 0 at `v j` for `i ≠ j`. -/
+def basis (v : ι ↪ F) : ι → F[X] := λ i, ∏ j, ite (i = j) 1 (basis_divisor v i j)
```
```suggestion
def basis (v : ι ↪ F) (i : ι) : F[X] := ∏ j in finset.univ.erase i, basis_divisor v i j
```

wrenna-robson replied:
This is quite a large change, and I'm not sure it's the right one. (Consider that `prod_ite` simplifies this to `finset.univ.filter` and not `finset.univ.erase`.)

wrenna-robson replied:
(I would be prepared to change it to `finset.univ.filter` though.)

linesthatinterlace committed:
Add bernoulli lemmas

linesthatinterlace committed:
Fix line lengths

linesthatinterlace committed:
Change definition of basis to use erase

linesthatinterlace committed:
Add characteristic property of the interpolation

linesthatinterlace committed:
Remove unused argument

linesthatinterlace committed:
Reposition module documentation

linesthatinterlace committed:
Add nodal/barycentric construction of Lagrange interpolant

linesthatinterlace committed:
Remove erase lemmas until we can figure out what to do with them

linesthatinterlace committed:
Fix line lengths

linesthatinterlace committed:
Fix line lengths

linesthatinterlace committed:
Refactor of previous commits to allow for finset indexing

linesthatinterlace committed:
Extend erase and sum of basis lemmas

linesthatinterlace committed:
Add barycentric form of interpolant

linesthatinterlace committed:
Remove Bernoulli lemmas as no longer used in this PR

BoltonBailey reviewed (commented):
This seems good. Scanning the proof no golfs jump out at me. I'll make some other reviews with style comments.

BoltonBailey commented on src/linear_algebra/lagrange.lean:
```diff
@@ -1,224 +1,540 @@
+theorem eq_zero_of_degree_lt_of_eval_index_eq_zero (hvs : set.inj_on v s)
+  (degree_f_lt : f.degree < s.card) (eval_f : ∀ i ∈ s, f.eval (v i) = 0) : f = 0 :=
+begin
+  classical, rw ← finset.card_image_of_inj_on hvs at degree_f_lt,
```
```suggestion
  classical, 
  rw ← finset.card_image_of_inj_on hvs at degree_f_lt,
```
There's something in the style guide about not putting multiple tactic invocations on one line unless you are closing a proof in one line.

BoltonBailey commented on src/linear_algebra/lagrange.lean:
```diff
@@ -1,224 +1,540 @@
 
+lemma sum_nodal_weight_mul_inv_sub_ne_zero (hvs : set.inj_on v s)
+  (hx : ∀ i ∈ s, x ≠ v i) (hs : s.nonempty)
+  : ∑ i in s, nodal_weight s v i * (x - v i)⁻¹ ≠ 0 :=
```
```suggestion
  (hx : ∀ i ∈ s, x ≠ v i) (hs : s.nonempty) :
  ∑ i in s, nodal_weight s v i * (x - v i)⁻¹ ≠ 0 :=
```
The style guide tells us to avoid putting a colon at the beginning of a line.

linesthatinterlace committed:
Apply suggestions from code review

Co-authored-by: Bolton Bailey <bolton.bailey@gmail.com>

Vierkantor commented on src/linear_algebra/lagrange.lean:
```diff
@@ -1,224 +1,541 @@
+  rw ← degree_lt_equiv_eq_zero_iff_eq_zero degree_f_lt,
+  exact matrix.eq_zero_of_forall_index_sum_mul_pow_eq_zero
+  (injective.comp (embedding.subtype _).inj' (equiv_fin_of_card_eq (card_coe _)).symm.injective)
+  (λ _, eval_f _ (finset.coe_mem _))
```
I'm OK with two tactic calls in exceptional cases if they express the same idea, but currently I find it hard to read with `rw` and `refine` all on the same line.
```suggestion
  rw ← mem_degree_lt at degree_f_lt,
  simp_rw eval_eq_sum_degree_lt_equiv degree_f_lt at eval_f,
  rw ← degree_lt_equiv_eq_zero_iff_eq_zero degree_f_lt,
  exact matrix.eq_zero_of_forall_index_sum_mul_pow_eq_zero
    (injective.comp (embedding.subtype _).inj' (equiv_fin_of_card_eq (card_coe _)).symm.injective)
    (λ _, eval_f _ (finset.coe_mem _))
```

Vierkantor commented on src/linear_algebra/lagrange.lean:
```diff
@@ -1,224 +1,541 @@
+theorem eq_of_degree_sub_lt_of_eval_eq (degree_fg_lt : (f - g).degree < s.card)
+  (eval_fg : ∀ x ∈ s, f.eval x = g.eval x) : f = g :=
+begin
+  rw ← sub_eq_zero, refine eq_zero_of_degree_lt_of_eval_eq_zero _ degree_fg_lt _,
```
```suggestion
  rw ← sub_eq_zero,
  refine eq_zero_of_degree_lt_of_eval_eq_zero _ degree_fg_lt _,
```

Vierkantor commented on src/linear_algebra/lagrange.lean:
```diff
@@ -1,224 +1,541 @@
+  (eval_fg : ∀ x ∈ s, f.eval x = g.eval x) : f = g :=
+begin
+  rw ← sub_eq_zero, refine eq_zero_of_degree_lt_of_eval_eq_zero _ degree_fg_lt _,
+  simp_rw [eval_sub, sub_eq_zero], exact eval_fg
```
```suggestion
  simp_rw [eval_sub, sub_eq_zero],
  exact eval_fg
```

Vierkantor commented on src/linear_algebra/lagrange.lean:
```diff
@@ -1,224 +1,541 @@
+  classical, 
+  rw ← finset.card_image_of_inj_on hvs at degree_f_lt,
+  refine eq_zero_of_degree_lt_of_eval_eq_zero _ degree_f_lt _,
+  refine λ x hx, _, rw finset.mem_image at hx, rcases hx with ⟨_, hj, rfl⟩, exact eval_f _ hj
```
```suggestion
  refine λ x hx, _,
  rw finset.mem_image at hx,
  rcases hx with ⟨_, hj, rfl⟩,
  exact eval_f _ hj
```

Vierkantor commented on src/linear_algebra/lagrange.lean:
```diff
@@ -1,224 +1,541 @@
+  f = g :=
+begin
+  rw ← sub_eq_zero, refine eq_zero_of_degree_lt_of_eval_index_eq_zero _ hvs degree_fg_lt _,
+  simp_rw [eval_sub, sub_eq_zero], exact eval_fg
```
```suggestion
  rw ← sub_eq_zero,
  refine eq_zero_of_degree_lt_of_eval_index_eq_zero _ hvs degree_fg_lt _,
  simp_rw [eval_sub, sub_eq_zero],
  exact eval_fg
```

Vierkantor commented on src/linear_algebra/lagrange.lean:
```diff
@@ -1,224 +1,541 @@
+begin
+  rw ← mem_degree_lt at degree_f_lt degree_g_lt,
+  refine eq_of_degree_sub_lt_of_eval_index_eq _ hvs _ eval_fg,
+  rw ← mem_degree_lt, exact submodule.sub_mem _ degree_f_lt degree_g_lt
```
```suggestion
  refine eq_of_degree_sub_lt_of_eval_index_eq _ hvs _ eval_fg,
  rw ← mem_degree_lt at degree_f_lt degree_g_lt ⊢,
  exact submodule.sub_mem _ degree_f_lt degree_g_lt
```

Vierkantor commented on src/linear_algebra/lagrange.lean:
```diff
@@ -1,224 +1,541 @@
+lemma basis_divisor_injective (hxy : basis_divisor x y = 0) : x = y :=
+begin
+  simp_rw [ basis_divisor, mul_eq_zero, X_sub_C_ne_zero, or_false,
+            C_eq_zero, inv_eq_zero, sub_eq_zero] at hxy, exact hxy
```
```suggestion
  simp_rw [basis_divisor, mul_eq_zero, X_sub_C_ne_zero, or_false,
            C_eq_zero, inv_eq_zero, sub_eq_zero] at hxy,
  exact hxy
```

Vierkantor commented on src/linear_algebra/lagrange.lean:
```diff
@@ -1,224 +1,541 @@
+by rw [basis_divisor_self_zero, degree_zero]
+
+lemma nat_degree_basis_divisor_eq : (basis_divisor x x).nat_degree = 0 :=
+by { rw [basis_divisor_self_zero, nat_degree_zero] }
```
```suggestion
by rw [basis_divisor_self_zero, nat_degree_zero]
```

Vierkantor commented on src/linear_algebra/lagrange.lean:
```diff
@@ -1,224 +1,541 @@
+@[simp] theorem basis_doubleton_left (hij : i ≠ j) :
+  basis {i, j} v i = basis_divisor (v i) (v j) :=
+by { simp only [  basis, hij, erase_insert_eq_erase, erase_eq_of_not_mem,
+                  mem_singleton, not_false_iff, prod_singleton] }
```
```suggestion
by { simp only [basis, hij, erase_insert_eq_erase, erase_eq_of_not_mem,
                mem_singleton, not_false_iff, prod_singleton] }
```

Vierkantor commented on src/linear_algebra/lagrange.lean:
```diff
@@ -1,224 +1,541 @@
+lemma basis_ne_zero (hvs : set.inj_on v s) (hi : i ∈ s) : basis s v i ≠ 0 :=
+begin
+  simp_rw [basis, prod_ne_zero_iff, ne.def, mem_erase], rintros j ⟨hij, hj⟩,
+  rw [basis_divisor_eq_zero_iff, hvs.eq_iff hi hj], exact hij.symm
```
```suggestion
  simp_rw [basis, prod_ne_zero_iff, ne.def, mem_erase],
  rintros j ⟨hij, hj⟩,
  rw [basis_divisor_eq_zero_iff, hvs.eq_iff hi hj],
  exact hij.symm
```

Vierkantor commented on src/linear_algebra/lagrange.lean:
```diff
@@ -1,224 +1,541 @@
+@[simp] theorem eval_basis_eq (hvs : set.inj_on v s) (hi : i ∈ s) : (basis s v i).eval (v i) = 1 :=
+begin
+  rw [basis, eval_prod], refine prod_eq_one (λ j H, _), rw [eval_basis_divisor_left_ne],
+  rw [mem_erase] at H, rcases H with ⟨hij, hj⟩, rw [ne.def, hvs.eq_iff hi hj], exact hij.symm
```
```suggestion
  rw [basis, eval_prod],
  refine prod_eq_one (λ j H, _),
  rw eval_basis_divisor_left_ne,
  rcases mem_erase.mp H with ⟨hij, hj⟩,
  exact mt (hvs hi hj) hij.symm
```

Vierkantor commented on src/linear_algebra/lagrange.lean:
```diff
@@ -1,224 +1,541 @@
+  rw [← card_erase_of_mem hi, card_eq_sum_ones],
+  convert nat_degree_prod _ _ H using 1,
+  refine sum_congr rfl (λ j hj, (nat_degree_basis_divisor_ne _).symm),
+  rw [ne.def, ← basis_divisor_eq_zero_iff], exact H _ hj
```
```suggestion
  have H : ∀ j, j ∈ s.erase i → basis_divisor (v i) (v j) ≠ 0,
  { simp_rw [ne.def, mem_erase, basis_divisor_eq_zero_iff],
    exact λ j ⟨hij₁, hj⟩ hij₂, hij₁ (hvs hj hi hij₂.symm) },
  rw [← card_erase_of_mem hi, card_eq_sum_ones],
  convert nat_degree_prod _ _ H using 1,
  refine sum_congr rfl (λ j hj, (nat_degree_basis_divisor_ne _).symm),
  rw [ne.def, ← basis_divisor_eq_zero_iff],
  exact H _ hj
```

Vierkantor commented on src/linear_algebra/lagrange.lean:
```diff
@@ -1,224 +1,541 @@
+  { intros i hi, rw eval_finset_sum, rw eval_one,
+    rw ← add_sum_erase _ _ hi, rw eval_basis_eq hvs hi, rw add_right_eq_self,
+    refine sum_eq_zero (λ j hj, _), rw mem_erase at hj, rcases hj with ⟨hij, hj⟩,
+    rw eval_basis_ne hij hi }
```
```suggestion
  refine eq_of_degrees_lt_of_eval_index_eq s hvs (lt_of_le_of_lt (degree_sum_le _ _) _) _ _,
  { rw finset.sup_lt_iff (with_bot.bot_lt_coe s.card),
    intros i hi,
    rw [degree_basis hvs hi, with_bot.coe_lt_coe],
    exact nat.pred_lt (card_ne_zero_of_mem hi) },
  { rw [degree_one, ← with_bot.coe_zero, with_bot.coe_lt_coe],
    exact nonempty.card_pos hs },
  { intros i hi,
    rw [eval_finset_sum, eval_one, ← add_sum_erase _ _ hi, eval_basis_eq hvs hi, add_right_eq_self],
    refine sum_eq_zero (λ j hj, _),
    rcases mem_erase.mp hj with ⟨hij, hj⟩,
    rw eval_basis_ne hij hi }
```

Vierkantor commented on src/linear_algebra/lagrange.lean:
```diff
@@ -1,224 +1,541 @@
+  classical, rw [ ←basis_sum (set.inj_on_of_injective function.injective_id _)
+                  ⟨x, mem_insert_self _ {y}⟩, sum_insert (not_mem_singleton.mpr hxy),
+                  sum_singleton, basis_doubleton_left hxy, basis_doubleton_right hxy, id, id]
+end
```
```suggestion
begin
  classical,
  rw [←basis_sum (set.inj_on_of_injective function.injective_id _) ⟨x, mem_insert_self _ {y}⟩,
      sum_insert (not_mem_singleton.mpr hxy), sum_singleton, basis_doubleton_left hxy,
      basis_doubleton_right hxy, id, id]
end
```

Vierkantor commented on src/linear_algebra/lagrange.lean:
```diff
@@ -1,224 +1,541 @@
-  { intro h, exact (h H).elim }
+  rw [interpolate, eval_finset_sum, ← add_sum_erase _ _ hi],
+  simp_rw [eval_mul, eval_C, eval_basis_eq hvs hi, mul_one, add_right_eq_self],
+  refine sum_eq_zero (λ j H, _), rw mem_erase at H, rw [eval_basis_ne H.1 hi, mul_zero]
```
```suggestion
  rw [interpolate_apply, eval_finset_sum, ← add_sum_erase _ _ hi],
  simp_rw [eval_mul, eval_C, eval_basis_eq hvs hi, mul_one, add_right_eq_self],
  refine sum_eq_zero (λ j H, _),
  rw [eval_basis_ne (mem_erase.mp H).1 hi, mul_zero]
```

Vierkantor commented on src/linear_algebra/lagrange.lean:
```diff
@@ -1,224 +1,541 @@
+by rw [degree_eq_nat_degree (basis_ne_zero hvs hi), nat_degree_basis hvs hi]
 
-@[simp] theorem nat_degree_basis (x : F) (hx : x ∈ s) : (basis s x).nat_degree = s.card - 1 :=
+lemma basis_sum (hvs : set.inj_on v s) (hs : s.nonempty) : ∑ j in s, (basis s v j) = 1 :=
```
The name `basis_sum` suggests `basis (Σ _)` to me, so:
```suggestion
lemma sum_basis (hvs : set.inj_on v s) (hs : s.nonempty) : ∑ j in s, (basis s v j) = 1 :=
```

Vierkantor commented on src/linear_algebra/lagrange.lean:
```diff
@@ -1,224 +1,541 @@
+  refine (degree_sum_le _ _).trans _, rw finset.sup_le_iff, intros i hi,
+  rw [degree_mul, degree_basis hvs hi], by_cases hr : (r i = 0),
+  { simp only [hr, map_zero, degree_zero, with_bot.bot_add], exact bot_le },
+  { rw [degree_C hr, zero_add, with_bot.coe_le_coe] }
```
```suggestion
  refine (degree_sum_le _ _).trans _,
  rw finset.sup_le_iff,
  intros i hi,
  rw [degree_mul, degree_basis hvs hi],
  by_cases hr : r i = 0,
  { simpa only [hr, map_zero, degree_zero, with_bot.bot_add] using bot_le },
  { rw [degree_C hr, zero_add, with_bot.coe_le_coe] }
```

Vierkantor commented on src/linear_algebra/lagrange.lean:
```diff
@@ -1,224 +1,541 @@
+  rcases eq_empty_or_nonempty s with rfl | h,
+  { rw [interpolate_empty, degree_zero, card_empty], exact with_bot.bot_lt_coe _ },
+  { refine lt_of_le_of_lt (degree_interpolate_le _ hvs) _, rw with_bot.coe_lt_coe,
+    exact nat.sub_lt (nonempty.card_pos h) (zero_lt_one) }
```
```suggestion
  rcases eq_empty_or_nonempty s with rfl | h,
  { rw [interpolate_empty, degree_zero, card_empty],
    exact with_bot.bot_lt_coe _ },
  { refine lt_of_le_of_lt (degree_interpolate_le _ hvs) _,
    rw with_bot.coe_lt_coe,
    exact nat.sub_lt (nonempty.card_pos h) zero_lt_one }
```

Vierkantor commented on src/linear_algebra/lagrange.lean:
```diff
@@ -1,224 +1,541 @@
+theorem values_eq_on_of_interpolate_eq (hvs : set.inj_on v s)
+  (hrr' : interpolate s v r = interpolate s v r') : ∀ i ∈ s, r i = r' i :=
+by { intros _ hi, rw [  ← eval_interpolate_at_node r hvs hi, hrr',
+                        eval_interpolate_at_node r' hvs hi] }
```
```suggestion
λ _ hi, by rw [← eval_interpolate_at_node r hvs hi, hrr', eval_interpolate_at_node r' hvs hi]
```

Vierkantor commented on src/linear_algebra/lagrange.lean:
```diff
@@ -1,224 +1,541 @@
+
+theorem interpolate_eq_of_values_eq_on (hrr' : ∀ i ∈ s, r i = r' i) :
+  interpolate s v r = interpolate s v r' :=
+by { simp_rw interpolate, refine sum_congr rfl (λ i hi, _), rw hrr' _ hi }
```
```suggestion
sum_congr rfl (λ i hi, (by rw hrr' _ hi))
```

Vierkantor commented on src/linear_algebra/lagrange.lean:
```diff
@@ -1,224 +1,541 @@
+by { simp_rw interpolate, refine sum_congr rfl (λ i hi, _), rw hrr' _ hi }
+
+@[simp] theorem interpolate_eq_iff_values_eq_on (hvs : set.inj_on v s)
+  : interpolate s v r = interpolate s v r' ↔ ∀ i ∈ s, r i = r' i :=
```
```suggestion
@[simp] theorem interpolate_eq_iff_values_eq_on (hvs : set.inj_on v s) :
  interpolate s v r = interpolate s v r' ↔ ∀ i ∈ s, r i = r' i :=
```

Vierkantor commented on src/linear_algebra/lagrange.lean:
```diff
@@ -1,224 +1,541 @@
+begin
+  refine eq_of_degrees_lt_of_eval_index_eq _ hvs degree_f_lt (degree_interpolate_lt _ hvs) _,
+  intros i hi, simp_rw [eval_interpolate_at_node _ hvs hi]
+end
```
```suggestion
eq_of_degrees_lt_of_eval_index_eq _ hvs degree_f_lt (degree_interpolate_lt _ hvs) $
λ i hi, (eval_interpolate_at_node _ hvs hi).symm
```

Vierkantor commented on src/linear_algebra/lagrange.lean:
```diff
@@ -1,224 +1,541 @@
+theorem eq_interpolate_of_eval_eq {f : F[X]} (hvs : set.inj_on v s)
+  (degree_f_lt : f.degree < s.card) (eval_f : ∀ i ∈ s, f.eval (v i) = r i) :
+  f = interpolate s v r :=
+by { rw [eq_interpolate hvs degree_f_lt], exact interpolate_eq_of_values_eq_on _ _ eval_f }
```
```suggestion
by { rw eq_interpolate hvs degree_f_lt, exact interpolate_eq_of_values_eq_on _ _ eval_f }
```

Vierkantor commented on src/linear_algebra/lagrange.lean:
```diff
@@ -1,224 +1,541 @@
+/-- Lagrange interpolation: given a finset `s : finset ι`, a nodal map  `v : ι → F`
+which is injective on s, and a value function `r : ι → F`, `interpolate x r` is the unique
+polynomial of degree `< s.card` that takes value `r i` on `v i` for all `i` in `s`. -/
+def interpolate (s : finset ι) (v r : ι → F) : F[X] := ∑ i in s, C (r i) * basis s v i
```
Since we don't need any extra properties of `interpolate` to prove that it's linear, we can just replace the definition of `interpolate` with the linear map (and get all the `interpolate_add` lemmas for free):
```suggestion
def interpolate (s : finset ι) (v : ι → F) : (ι → F) →ₗ[F] F[X] :=
{ to_fun := λ r, ∑ i in s, C (r i) * basis s v i,
  map_add' := λ f g, by simp_rw [← finset.sum_add_distrib, ← add_mul,
                                 ← C_add, pi.add_apply],
  map_smul' := λ c f, by simp_rw [finset.smul_sum, C_mul', smul_smul,
                                  pi.smul_apply, ring_hom.id_apply, smul_eq_mul] }

lemma interpolate_apply (s : finset ι) (v r : ι → F) :
  interpolate s v r = ∑ i in s, C (r i) * basis s v i :=
rfl
```

Vierkantor commented on src/linear_algebra/lagrange.lean:
```diff
@@ -1,224 +1,541 @@
 
-@[simp] theorem interpolate_empty (f) : interpolate (∅ : finset F) f = 0 :=
-rfl
+@[simp] theorem interpolate_empty : interpolate ∅ v r = 0 := sum_empty
```
```suggestion
@[simp] theorem interpolate_empty : interpolate ∅ v r = 0 :=
by rw [interpolate_apply, sum_empty]
```

Vierkantor commented on src/linear_algebra/lagrange.lean:
```diff
@@ -1,224 +1,541 @@
-@[simp] theorem interpolate_singleton (f) (x : F) : interpolate {x} f = C (f x) :=
-by rw [interpolate, finset.sum_singleton, basis_singleton_self, mul_one]
+@[simp] theorem interpolate_singleton : interpolate {i} v r = C (r i) :=
+by rw [interpolate, sum_singleton, basis_singleton, mul_one]
```
```suggestion
by rw [interpolate_apply, sum_singleton, basis_singleton, mul_one]
```

Vierkantor commented on src/linear_algebra/lagrange.lean:
```diff
@@ -1,224 +1,541 @@
 
-@[simp] theorem eval_interpolate (x) (H : x ∈ s) : eval x (interpolate s f) = f x :=
+@[simp] theorem interpolate_one (hvs : set.inj_on v s) (hs : s.nonempty) : interpolate s v 1 = 1 :=
+by { simp_rw [interpolate, pi.one_apply, map_one, one_mul], exact basis_sum hvs hs }
```
```suggestion
by { simp_rw [interpolate_apply, pi.one_apply, map_one, one_mul], exact basis_sum hvs hs }
```

Vierkantor commented on src/linear_algebra/lagrange.lean:
```diff
@@ -1,224 +1,541 @@
+{ to_fun := interpolate s v,
+  map_add' := λ f g, by { simp_rw [ interpolate, ← finset.sum_add_distrib,
+                                    ← add_mul, ← C_add], refl },
   map_smul' := λ c f, by { simp_rw [interpolate, finset.smul_sum, C_mul', smul_smul], refl } }
```
If we define `interpolate` as a linear map, we can delete all this:
```suggestion
```

Vierkantor commented on src/linear_algebra/lagrange.lean:
```diff
@@ -1,224 +1,541 @@
+  { rw h, exact ⟨degree_interpolate_lt _ hvs, λ _ hi, eval_interpolate_at_node _ hvs hi⟩ }
 end
 
 /-- Linear version of `interpolate`. -/
```
```suggestion
```

Vierkantor commented on src/linear_algebra/lagrange.lean:
```diff
@@ -1,224 +1,541 @@
+(linterpolate s v).map_sub r r'
+
+@[simp] lemma interpolate_smul (c : F) (r) : interpolate s v (c • r) = c • interpolate s v r :=
+(linterpolate s v).map_smul c r
```
```suggestion
```

Vierkantor commented on src/linear_algebra/lagrange.lean:
```diff
@@ -1,224 +1,541 @@
+  right_inv :=  begin
+                  intro f, ext i, rcases i with ⟨i, hi⟩, simp only [subtype.coe_mk],
+                  rw eval_interpolate_at_node _ hvs hi, exact dif_pos hi,
+                end }
```
```suggestion
{ to_fun := λ f i, f.1.eval (v i),
  map_add' := λ f g, funext $ λ v, eval_add,
  map_smul' := λ c f, funext $ by simp,
  inv_fun := λ r, ⟨interpolate s v (λ x, if hx : x ∈ s then r ⟨x, hx⟩ else 0),
                   mem_degree_lt.2 $ degree_interpolate_lt _ hvs⟩,
  left_inv := begin
    rintros ⟨f, hf⟩,
    simp only [subtype.mk_eq_mk, subtype.coe_mk, dite_eq_ite],
    rw mem_degree_lt at hf,
    nth_rewrite_rhs 0 eq_interpolate hvs hf,
    rw interpolate_eq_iff_values_eq_on _ _ hvs,
    exact λ _ hi, if_pos hi
  end,
  right_inv :=
  begin
    intro f,
    ext ⟨i, hi⟩,
    simp only [subtype.coe_mk, eval_interpolate_at_node _ hvs hi],
    exact dif_pos hi,
  end }
```

Vierkantor commented on src/linear_algebra/lagrange.lean:
```diff
@@ -1,224 +1,541 @@
+      refine sum_congr rfl (λ j hj, _), congr,
+      refine eval_interpolate_at_node _ (hvt.mono (insert_subset.mpr ⟨hst hj, sdiff_subset _ _⟩))
+                                (mem_insert .mpr (or.inr (mem_sdiff.mpr ⟨hi, hi'⟩))) } }
+end
```
```suggestion
begin
  symmetry,
  refine eq_interpolate_of_eval_eq _ hvt (lt_of_le_of_lt (degree_sum_le _ _) _) (λ i hi, _),
  { simp_rw [(finset.sup_lt_iff (with_bot.bot_lt_coe t.card)), degree_mul],
    intros i hi,
    have hs : 1 ≤ s.card := nonempty.card_pos ⟨_, hi⟩,
    have hst' : s.card ≤ t.card := card_le_of_subset hst,
    have H : t.card = (1 + (t.card - s.card)) + (s.card - 1),
    { rw [add_assoc, tsub_add_tsub_cancel hst' hs, ← add_tsub_assoc_of_le (hs.trans hst'),
          nat.succ_add_sub_one, zero_add] },

    rw [degree_basis (set.inj_on.mono hst hvt) hi, H, with_bot.coe_add,
        with_bot.add_lt_add_iff_right (@with_bot.coe_ne_bot _ (s.card - 1))],
    convert degree_interpolate_lt _ (hvt.mono (coe_subset.mpr (insert_subset.mpr
      ⟨hst hi, sdiff_subset _ _⟩))),
    rw [card_insert_of_not_mem (not_mem_sdiff_of_mem_right hi), card_sdiff hst, add_comm] },

  { simp_rw [eval_finset_sum, eval_mul],
    by_cases hi' : i ∈ s,
    { rw [← add_sum_erase _ _ hi', eval_basis_eq (hvt.mono hst) hi',
          eval_interpolate_at_node _ (hvt.mono (coe_subset.mpr
            (insert_subset.mpr ⟨hi, sdiff_subset _ _⟩))) (mem_insert_self _ _),
          mul_one, add_right_eq_self],
      refine sum_eq_zero (λ j hj, _),
      rcases mem_erase.mp hj with ⟨hij, hj⟩,
      rw [eval_basis_ne hij hi', mul_zero] },
    { have H : ∑ j in s, eval (v i) (basis s v j) = 1,
      { rw [← eval_finset_sum, basis_sum (hvt.mono hst) hs, eval_one] },
      rw [← mul_one (r i), ← H, mul_sum],
      refine sum_congr rfl (λ j hj, _),
      congr,
      exact eval_interpolate_at_node _ (hvt.mono (insert_subset.mpr ⟨hst hj, sdiff_subset _ _⟩))
                                (mem_insert.mpr (or.inr (mem_sdiff.mpr ⟨hi, hi'⟩))) } }
end
```

Vierkantor commented on src/linear_algebra/lagrange.lean:
```diff
@@ -1,224 +1,541 @@
+  push_neg; exact ⟨ by { rintros (rfl | ⟨_, _, _⟩); tauto },
+                    by { rw or_iff_not_imp_left, rintros ⟨_, _⟩ _, tauto, } ⟩,
+  intros k hk, simp at hk, rcases hk with (rfl | rfl), exact hi, exact hj
+end
```
Here's a more principled proof, but it needs the following `finset` lemmas. Could you add those to `data/finset/basic.lean`?

```lean
lemma finset.sdiff_sdiff (s t u : finset ι) :
  (s \ t) \ u = (s \ t) ∩ (s \ u) :=
by ext; finish

lemma finset.sdiff_insert (s t : finset ι) (x : ι) :
  s \ insert x t = (s \ t).erase x :=
by simp only [← sdiff_singleton_eq_erase, insert_eq, sdiff_sdiff, sdiff_union_distrib, inter_comm]

lemma finset.insert_sdiff_insert {s t : finset ι} {x : ι} (hxs : x ∈ s) (hxt : x ∉ t) :
  insert x (s \ insert x t) = s \ t :=
by rw [sdiff_insert, insert_erase (mem_sdiff.mpr ⟨hxs, hxt⟩)]

lemma finset.doubleton_comm (x y : ι) : ({x, y} : finset ι) = {y, x} :=
by ext; finish
```

```suggestion
begin
  rw [interpolate_eq_sum_interpolate_insert_sdiff _ hvs ⟨i, (mem_insert_self i {j})⟩ _,
      sum_insert (not_mem_singleton.mpr hij), sum_singleton,
      basis_doubleton_left hij, basis_doubleton_right hij,
      finset.insert_sdiff_insert hi (not_mem_singleton.mpr hij), sdiff_singleton_eq_erase,
      finset.doubleton_comm, finset.insert_sdiff_insert hj (not_mem_singleton.mpr hij.symm),
      sdiff_singleton_eq_erase],
  { exact insert_subset.mpr ⟨hi, singleton_subset_iff.mpr hj⟩ },
end
```

wrenna-robson replied:
I've added this. Note that the sdiff lemmas in finset, compared to the equivalent lemmas in generalized boolean algebra, seem piecemeal and a little bit of a mess in terms of naming. I haven't tried to fix that here but I just want to note that that could be a good future PR.

Vierkantor commented on src/linear_algebra/lagrange.lean:
```diff
@@ -1,224 +1,541 @@
-... ≤ f.roots.card : with_bot.coe_le_coe.2 $ f.roots.to_finset_card_le
-... ≤ f.degree : card_roots hf3
+lemma X_sub_C_dvd_nodal (v : ι → F) (hi : i ∈ s) : (X - C (v i)) ∣ nodal s v :=
+⟨_, by { classical, exact nodal_eq_mul_nodal_erase hi}⟩
```
```suggestion
⟨_, by { classical, exact nodal_eq_mul_nodal_erase hi }⟩
```

Vierkantor commented on src/linear_algebra/lagrange.lean:
```diff
@@ -1,224 +1,541 @@
+lemma nodal_erase_eq_nodal_div (hi : i ∈ s) :
+  nodal (s.erase i) v = nodal s v / (X - C (v i)) :=
+by {  rw [nodal_eq_mul_nodal_erase hi, euclidean_domain.mul_div_cancel_left],
+      exact X_sub_C_ne_zero _ }
```
```suggestion
begin
  rw [nodal_eq_mul_nodal_erase hi, euclidean_domain.mul_div_cancel_left],
  exact X_sub_C_ne_zero _
end
```

Vierkantor commented on src/linear_algebra/lagrange.lean:
```diff
@@ -1,224 +1,541 @@
+
+lemma derivative_nodal : (nodal s v).derivative = ∑ i in s, nodal (s.erase i) v :=
+begin
+refine finset.induction_on s _ (λ _ _ hit IH, _),
```
```suggestion
  refine finset.induction_on s _ (λ _ _ hit IH, _),
```

Vierkantor commented on src/linear_algebra/lagrange.lean:
```diff
@@ -1,224 +1,541 @@
+    refine sum_congr rfl (λ j hjt, _),
+    rw [  nodal_erase_eq_nodal_div (mem_insert_of_mem hjt), nodal_insert_eq_nodal hit,
+          euclidean_domain.mul_div_assoc _ (X_sub_C_dvd_nodal v hjt),
+          nodal_erase_eq_nodal_div hjt] }
```
```suggestion
  { rw [nodal_insert_eq_nodal hit, derivative_mul, IH, derivative_sub,
        derivative_X, derivative_C, sub_zero, one_mul, sum_insert hit,
        mul_sum, erase_insert hit, add_right_inj],
    refine sum_congr rfl (λ j hjt, _),
    rw [nodal_erase_eq_nodal_div (mem_insert_of_mem hjt), nodal_insert_eq_nodal hit,
        euclidean_domain.mul_div_assoc _ (X_sub_C_dvd_nodal v hjt),
        nodal_erase_eq_nodal_div hjt] }
```

Vierkantor commented on src/linear_algebra/lagrange.lean:
```diff
@@ -1,224 +1,541 @@
+  rw [derivative_nodal, eval_finset_sum, ← add_sum_erase _ _ hi, add_right_eq_self],
+  refine sum_eq_zero (λ j hj, _),
+  simp_rw [nodal, eval_prod, eval_sub, eval_X, eval_C, prod_eq_zero_iff, mem_erase],
+  rw mem_erase at hj, refine ⟨i, ⟨hj.1.symm, hi⟩, sub_eq_zero_of_eq rfl⟩
```
```suggestion
  rw [derivative_nodal, eval_finset_sum, ← add_sum_erase _ _ hi, add_right_eq_self],
  refine sum_eq_zero (λ j hj, _),
  simp_rw [nodal, eval_prod, eval_sub, eval_X, eval_C, prod_eq_zero_iff, mem_erase],
  exact ⟨i, ⟨(mem_erase.mp hj).1.symm, hi⟩, sub_eq_zero_of_eq rfl⟩
```

Vierkantor commented on src/linear_algebra/lagrange.lean:
```diff
@@ -1,224 +1,541 @@
+begin
+  rw nodal_weight, rw prod_ne_zero_iff, intros j hj, rw mem_erase at hj, rcases hj with ⟨hij, hj⟩,
+  refine inv_ne_zero (sub_ne_zero_of_ne _), rw [ne.def, hvs.eq_iff hi hj], exact hij.symm
+end
```
```suggestion
begin
  rw [nodal_weight, prod_ne_zero_iff],
  intros j hj,
  rcases mem_erase.mp hj with ⟨hij, hj⟩,
  refine inv_ne_zero (sub_ne_zero_of_ne (mt (hvs.eq_iff hi hj).mp hij.symm)),
end
```

Vierkantor commented on src/linear_algebra/lagrange.lean:
```diff
@@ -1,224 +1,541 @@
+lemma basis_eq_prod_sub_inv_mul_nodal_div (hi : i ∈ s) :
+  basis s v i = C (nodal_weight s v i) * ( nodal s v / (X - C (v i)) )  :=
+by {  simp_rw [ basis, basis_divisor, nodal_weight, prod_mul_distrib,
+                map_prod, ← nodal_erase_eq_nodal_div hi, nodal]  }
```
```suggestion
by simp_rw [basis, basis_divisor, nodal_weight, prod_mul_distrib,
            map_prod, ← nodal_erase_eq_nodal_div hi, nodal]
```

Vierkantor commented on src/linear_algebra/lagrange.lean:
```diff
@@ -1,224 +1,541 @@
+  eval x (basis s v i) = (eval x (nodal s v)) * (nodal_weight s v i * (x - v i)⁻¹)  :=
+by rw [ mul_comm, basis_eq_prod_sub_inv_mul_nodal_div hi, eval_mul, eval_C,
+        ← nodal_erase_eq_nodal_div hi, eval_nodal, eval_nodal, mul_assoc, ← mul_prod_erase _ _ hi,
+        ← mul_assoc (x - v i)⁻¹, inv_mul_cancel (sub_ne_zero_of_ne hxi), one_mul]
```
```suggestion
by rw [mul_comm, basis_eq_prod_sub_inv_mul_nodal_div hi, eval_mul, eval_C,
       ← nodal_erase_eq_nodal_div hi, eval_nodal, eval_nodal, mul_assoc, ← mul_prod_erase _ _ hi,
       ← mul_assoc (x - v i)⁻¹, inv_mul_cancel (sub_ne_zero_of_ne hxi), one_mul]
```

Vierkantor commented on src/linear_algebra/lagrange.lean:
```diff
@@ -1,224 +1,541 @@
+
+lemma interpolate_eq_nodal_weight_mul_nodal_div_X_sub_C :
+interpolate s v r = ∑ i in s, C (nodal_weight s v i) * (nodal s v / (X - C (v i))) * C (r i) :=
+by {  refine sum_congr rfl (λ j hj, _), rw [mul_comm, basis_eq_prod_sub_inv_mul_nodal_div hj]  }
```
```suggestion
sum_congr rfl (λ j hj, by rw [mul_comm, basis_eq_prod_sub_inv_mul_nodal_div hj])
```

Vierkantor commented on src/linear_algebra/lagrange.lean:
```diff
@@ -1,224 +1,541 @@
+  refine @right_ne_zero_of_mul_eq_one  _ _ _ (eval x (nodal s v)) _ _,
+  have H := eval_interpolate_not_at_node 1 hx,
+  dsimp at H, rw interpolate_one hvs hs at H, rw eval_one at H, rw H, simp_rw mul_one,
+end
```
```suggestion
@right_ne_zero_of_mul_eq_one  _ _ _ (eval x (nodal s v)) _ $
  by simpa only [pi.one_apply, interpolate_one hvs hs, eval_one, mul_one]
    using (eval_interpolate_not_at_node 1 hx).symm
```

Vierkantor commented on src/linear_algebra/lagrange.lean:
```diff
@@ -1,224 +1,541 @@
+  ∑ i in s, nodal_weight s v i * (x - v i)⁻¹ :=
+by { rw [  ← div_one (eval x (interpolate s v r)), ← @eval_one _ _ x, ← interpolate_one hvs hs,
+        eval_interpolate_not_at_node r hx, eval_interpolate_not_at_node 1 hx],
+  simp_rw [mul_div_mul_left _ _ (eval_nodal_not_at_node hx), pi.one_apply, mul_one] }
```
```suggestion
begin
  rw [← div_one (eval x (interpolate s v r)), ← @eval_one _ _ x, ← interpolate_one hvs hs,
      eval_interpolate_not_at_node r hx, eval_interpolate_not_at_node 1 hx],
  simp only [mul_div_mul_left _ _ (eval_nodal_not_at_node hx), pi.one_apply, mul_one]
end
```

Vierkantor commented on src/linear_algebra/vandermonde.lean:
```diff
@@ -133,4 +133,19 @@ lemma det_vandermonde_ne_zero_iff [is_domain R] {n : ℕ} {v : fin n → R} :
+theorem eq_zero_of_forall_pow_sum_mul_pow_eq_zero {R : Type*} [comm_ring R]
+[is_domain R] {n : ℕ} {f v : fin n → R} (hf : function.injective f)
+(hfv : ∀ i : fin n, ∑ j : fin n, v j * (f j ^ (i : ℕ)) = 0) : v = 0
+:= eq_zero_of_vec_mul_eq_zero (det_vandermonde_ne_zero_iff.mpr hf) (funext hfv)
```
```suggestion
theorem eq_zero_of_forall_index_sum_pow_mul_eq_zero {R : Type*} [comm_ring R]
  [is_domain R] {n : ℕ} {f v : fin n → R} (hf : function.injective f)
  (hfv : ∀ j, ∑ i : fin n, (f j ^ (i : ℕ)) * v i = 0) : v = 0 :=
eq_zero_of_mul_vec_eq_zero (det_vandermonde_ne_zero_iff.mpr hf) (funext hfv)

theorem eq_zero_of_forall_index_sum_mul_pow_eq_zero {R : Type*} [comm_ring R]
  [is_domain R] {n : ℕ} {f v : fin n → R} (hf : function.injective f)
  (hfv : ∀ j, ∑ i, v i * (f j ^ (i : ℕ)) = 0) : v = 0 :=
by { refine eq_zero_of_forall_index_sum_pow_mul_eq_zero hf _, simp_rw mul_comm, exact hfv }

theorem eq_zero_of_forall_pow_sum_mul_pow_eq_zero {R : Type*} [comm_ring R]
  [is_domain R] {n : ℕ} {f v : fin n → R} (hf : function.injective f)
  (hfv : ∀ i : fin n, ∑ j : fin n, v j * (f j ^ (i : ℕ)) = 0) : v = 0 :=
eq_zero_of_vec_mul_eq_zero (det_vandermonde_ne_zero_iff.mpr hf) (funext hfv)
```

Vierkantor reviewed (commented):
I've got a lot of formatting remarks, apart from that the structure looks good overall. I'm just wondering whether there's a nicer way to deal with all the `mem_erase` rewriting going on, but haven't found a better approach.

Vierkantor commented on src/linear_algebra/lagrange.lean:
```diff
@@ -1,224 +1,541 @@
 import algebra.big_operators.basic
 import ring_theory.polynomial.basic
+import logic.lemmas
+import linear_algebra.vandermonde
```
```suggestion
import algebra.big_operators.basic
import linear_algebra.vandermonde
import logic.lemmas
import ring_theory.polynomial.basic
```

linesthatinterlace committed:
Apply suggestions from code review

Co-authored-by: Anne Baanen <Vierkantor@users.noreply.github.com>

linesthatinterlace committed:
Fix issues arising from code review changes

linesthatinterlace committed:
Improve add_erase proof

Vierkantor commented on src/data/finset/basic.lean:
```diff
@@ -1232,6 +1235,18 @@ by { ext, rw [mem_erase, mem_sdiff, mem_singleton], tauto }
+lemma sdiff_insert (s t : finset α) (x : α) :
+  s \ insert x t = (s \ t).erase x :=
+by simp_rw [  ← sdiff_singleton_eq_erase, insert_eq,
+              sdiff_sdiff_left', sdiff_union_distrib, inter_comm]
```
```suggestion
by simp_rw [← sdiff_singleton_eq_erase, insert_eq,
            sdiff_sdiff_left', sdiff_union_distrib, inter_comm]
```

Vierkantor commented on src/ring_theory/polynomial/basic.lean:
```diff
@@ -131,6 +131,17 @@ def degree_lt_equiv (R) [semiring R] (n : ℕ) : degree_lt R n ≃ₗ[R] (fin n
 
+@[simp] theorem degree_lt_equiv_eq_zero_iff_eq_zero {n : ℕ} {p : R[X]} (hp : p ∈ degree_lt R n) :
+  degree_lt_equiv _ _ ⟨p, hp⟩ = 0 ↔ p = 0 :=
+by {rw [linear_equiv.map_eq_zero_iff, submodule.mk_eq_zero]}
```
```suggestion
by { rw [linear_equiv.map_eq_zero_iff, submodule.mk_eq_zero] }
```

vihdzp replied:
How about this instead?
```suggestion
by rw [linear_equiv.map_eq_zero_iff, submodule.mk_eq_zero]
```

Vierkantor commented on src/linear_algebra/lagrange.lean:
```diff
@@ -1,224 +1,567 @@
+open polynomial
 
-universe u
+section polynomial_determination
```
This section should all go in the `polynomial` namespace.
```suggestion
section polynomial_determination

namespace polynomial
```

Vierkantor commented on src/linear_algebra/lagrange.lean:
```diff
@@ -1,224 +1,567 @@
-@[simp] theorem nat_degree_basis (x : F) (hx : x ∈ s) : (basis s x).nat_degree = s.card - 1 :=
+end indexed
+
+end polynomial_determination
```
```suggestion
end polynomial

end polynomial_determination
```

Vierkantor commented on src/linear_algebra/lagrange.lean:
```diff
@@ -1,224 +1,567 @@
+variables (s : finset R)
 
-open polynomial
+theorem eq_zero_of_degree_lt_of_eval_eq_zero (degree_f_lt : f.degree < s.card)
```
I'm not so convinced by the name, it feels like it's missing a reference to `finset`. How about:
```suggestion
theorem eq_zero_of_degree_lt_of_eval_finset_eq_zero (degree_f_lt : f.degree < s.card)
```

Vierkantor commented on src/linear_algebra/lagrange.lean:
```diff
@@ -1,224 +1,567 @@
-/-- Lagrange basis polynomials that evaluate to 1 at `x` and 0 at other elements of `s`. -/
-def basis (x : F) : F[X] :=
-∏ y in s.erase x, C (x - y)⁻¹ * (X - C y)
+theorem eq_of_degree_sub_lt_of_eval_eq (degree_fg_lt : (f - g).degree < s.card)
```
```suggestion
theorem eq_of_degree_sub_lt_of_eval_finset_eq (degree_fg_lt : (f - g).degree < s.card)
```

Vierkantor commented on src/linear_algebra/lagrange.lean:
```diff
@@ -1,224 +1,567 @@
 
-@[simp] theorem basis_empty (x : F) : basis ∅ x = 1 :=
-rfl
+theorem eq_of_degrees_lt_of_eval_eq (degree_f_lt : f.degree < s.card)
```
```suggestion
theorem eq_of_degrees_lt_of_eval_finset_eq (degree_f_lt : f.degree < s.card)
```

Vierkantor commented on src/linear_algebra/lagrange.lean:
```diff
@@ -1,224 +1,567 @@
+  classical,
+  rw ← finset.card_image_of_inj_on hvs at degree_f_lt,
+  refine eq_zero_of_degree_lt_of_eval_eq_zero _ degree_f_lt _,
+  refine λ x hx, _,
```
Why not just:
```suggestion
  intros x hx,
```

Vierkantor commented on src/linear_algebra/lagrange.lean:
```diff
@@ -1,224 +1,567 @@
+  refine eq_zero_of_degree_lt_of_eval_eq_zero _ degree_f_lt _,
+  refine λ x hx, _,
+  rw finset.mem_image at hx,
+  rcases hx with ⟨_, hj, rfl⟩,
```
```suggestion
  rcases finset.mem_image.mp hx with ⟨_, hj, rfl⟩,
```

Vierkantor commented on src/linear_algebra/lagrange.lean:
```diff
@@ -1,224 +1,567 @@
+lemma basis_divisor_ne_zero_iff : basis_divisor x y ≠ 0 ↔ x ≠ y :=
+by rw [ne.def, basis_divisor_eq_zero_iff]
+
+lemma degree_basis_divisor_ne (hxy : x ≠ y) : (basis_divisor x y).degree = 1 :=
```
I'd like to see the fact that the `ne` is in the hypothesis, not the conclusion (although there are counterexamples: [`lp.single_apply_ne`](https://leanprover-community.github.io/mathlib_docs/analysis/normed_space/lp_space.html#lp.single_apply_ne), `of_ne` is a lot more common in my experience).
```suggestion
lemma degree_basis_divisor_of_ne (hxy : x ≠ y) : (basis_divisor x y).degree = 1 :=
```

Vierkantor commented on src/linear_algebra/lagrange.lean:
```diff
@@ -1,224 +1,567 @@
+lemma basis_divisor_self_zero : basis_divisor x x = 0 :=
+by simp only [basis_divisor, sub_self, inv_zero, map_zero, zero_mul]
+
+lemma basis_divisor_injective (hxy : basis_divisor x y = 0) : x = y :=
```
Usually we restrict `_injective` for a statement of `function.injective`
```suggestion
lemma basis_divisor_inj (hxy : basis_divisor x y = 0) : x = y :=
```

Vierkantor commented on src/linear_algebra/lagrange.lean:
```diff
@@ -1,224 +1,567 @@
+Conceptually, they are the building blocks for the Lagrange interpolants. -/
+def basis_divisor (x y : F) : F[X] := C ((x - y)⁻¹) * (X - C (y))
+
+lemma basis_divisor_self_zero : basis_divisor x x = 0 :=
```
```suggestion
lemma basis_divisor_self : basis_divisor x x = 0 :=
```

Vierkantor commented on src/linear_algebra/lagrange.lean:
```diff
@@ -1,224 +1,567 @@
-that takes value `f x` on all `x` in `s`. -/
-def interpolate : F[X] :=
-∑ x in s, C (f x) * basis s x
+@[simp] lemma degree_basis_divisor_eq : (basis_divisor x x).degree = ⊥ :=
```
```suggestion
@[simp] lemma degree_basis_divisor_self : (basis_divisor x x).degree = ⊥ :=
```

Vierkantor commented on src/linear_algebra/lagrange.lean:
```diff
@@ -1,224 +1,567 @@
 
-@[simp] theorem interpolate_empty (f) : interpolate (∅ : finset F) f = 0 :=
-rfl
+lemma nat_degree_basis_divisor_eq : (basis_divisor x x).nat_degree = 0 :=
```
For consistency with `basis_divisor_self`(`_zero`):
```suggestion
lemma nat_degree_basis_divisor_self : (basis_divisor x x).nat_degree = 0 :=
```

Vierkantor commented on src/linear_algebra/lagrange.lean:
```diff
@@ -1,224 +1,567 @@
 
-@[simp] theorem interpolate_singleton (f) (x : F) : interpolate {x} f = C (f x) :=
-by rw [interpolate, finset.sum_singleton, basis_singleton_self, mul_one]
+lemma nat_degree_basis_divisor_ne (hxy : x ≠ y) : (basis_divisor x y).nat_degree = 1 :=
```
```suggestion
lemma nat_degree_basis_divisor_of_ne (hxy : x ≠ y) : (basis_divisor x y).nat_degree = 1 :=
```

Vierkantor commented on src/linear_algebra/lagrange.lean:
```diff
@@ -1,224 +1,567 @@
+@[simp] lemma eval_basis_divisor_right : eval y (basis_divisor x y) = 0 :=
+by simp only [basis_divisor, eval_mul, eval_C, eval_sub, eval_X, sub_self, mul_zero]
+
+lemma eval_basis_divisor_left_ne (hxy : x ≠ y) : eval x (basis_divisor x y) = 1 :=
```
```suggestion
lemma eval_basis_divisor_left_of_ne (hxy : x ≠ y) : eval x (basis_divisor x y) = 1 :=
```

Vierkantor commented on src/linear_algebra/lagrange.lean:
```diff
@@ -1,224 +1,567 @@
+  exact hij.symm
+end
+
+@[simp] theorem eval_basis_eq (hvs : set.inj_on v s) (hi : i ∈ s) : (basis s v i).eval (v i) = 1 :=
```
```suggestion
@[simp] theorem eval_basis_self (hvs : set.inj_on v s) (hi : i ∈ s) : (basis s v i).eval (v i) = 1 :=
```

Vierkantor commented on src/linear_algebra/lagrange.lean:
```diff
@@ -1,224 +1,567 @@
+/-- Lagrange basis polynomials indexed by `s : finset ι`,
+defined at nodes `v i` for a map `v : ι → F` For `i, j ∈ s`, `basis s v i` evaluates to 1 at
+`v i` and 0 at `v j` for `i ≠ j`. -/
+def basis (s : finset ι) (v : ι → F) (i : ι) : F[X] := ∏ j in s.erase i, basis_divisor (v i) (v j)
```
This is probably going to conflict with `basis` from linear algebra, so unless it gives too much noise let's make it `protected`.
```suggestion
protected def basis (s : finset ι) (v : ι → F) (i : ι) : F[X] := ∏ j in s.erase i, basis_divisor (v i) (v j)
```

Vierkantor commented on src/linear_algebra/lagrange.lean:
```diff
@@ -1,224 +1,567 @@
-... ≤ (basis s b).nat_degree : degree_le_nat_degree
-... = (s.card - 1 : ℕ) : by { rwa nat_degree_basis }
-... < s.card : with_bot.coe_lt_coe.2 (nat.pred_lt $ mt finset.card_eq_zero.1 H)
+@[simp] theorem eval_basis_ne (hij : i ≠ j) (hj : j ∈ s) : (basis s v i).eval (v j) = 0 :=
```
```suggestion
@[simp] theorem eval_basis_of_ne (hij : i ≠ j) (hj : j ∈ s) : (basis s v i).eval (v j) = 0 :=
```

Vierkantor commented on src/linear_algebra/lagrange.lean:
```diff
@@ -1,224 +1,567 @@
+/-- Lagrange interpolation: given a finset `s : finset ι`, a nodal map  `v : ι → F`
+which is injective on s, and a value function `r : ι → F`, `interpolate x r` is the unique
+polynomial of degree `< s.card` that takes value `r i` on `v i` for all `i` in `s`. -/
+def interpolate (s : finset ι) (v : ι → F) : (ι → F) →ₗ[F] F[X] :=
```
The `simps` attribute should be able to automatically generate `interpolate_apply`:
```suggestion
@[simps]
def interpolate (s : finset ι) (v : ι → F) : (ι → F) →ₗ[F] F[X] :=
```

Vierkantor commented on src/linear_algebra/lagrange.lean:
```diff
@@ -1,224 +1,567 @@
+variables {ι : Type*} {s : finset ι} {v : ι → F} {i : ι} (r : ι → F) {x : F}
+
+/--
+We define `nodal`, the unique monic polynomial whose roots are the nodes defined by `v` and `s`.
```
I didn't immediately understand this use of "node", maybe add an explanation in plainer words?
```suggestion
`nodal s v` is the unique monic polynomial whose roots are the nodes defined by `v` and `s`.

That is, the roots of `nodal s v` are exactly the image of `v` on `s`, with appropriate multiplicity.
```

Vierkantor reviewed (commented):
I went through again especially focussed on the names this time, I think one more round of reviews and we're there.

vihdzp commented on src/data/finset/basic.lean:
```diff
@@ -795,6 +795,9 @@ by simp only [insert_eq, union_left_comm]
 lemma insert_union_distrib (a : α) (s t : finset α) : insert a (s ∪ t) = insert a s ∪ insert a t :=
 by simp only [insert_union, union_insert, insert_idem]
 
+lemma doubleton_comm (x y : α) : ({x, y} : finset α) = {y, x} :=
```
This already exists as `finset.pair_comm`. Please see the comment just below [`has_ssubset`](https://leanprover-community.github.io/mathlib_docs/init/core.html#has_ssubset).

wrenna-robson replied:
Then we should change the name of `finset.card_doubleton` and `list.doubleton_eq`.

eric-wieser replied:
I added this to our naming guide, https://github.com/leanprover-community/leanprover-community.github.io/pull/276

vihdzp commented on src/linear_algebra/vandermonde.lean:
```diff
@@ -133,4 +133,19 @@ lemma det_vandermonde_ne_zero_iff [is_domain R] {n : ℕ} {v : fin n → R} :
+theorem eq_zero_of_forall_index_sum_mul_pow_eq_zero {R : Type*} [comm_ring R]
+  [is_domain R] {n : ℕ} {f v : fin n → R} (hf : function.injective f)
+  (hfv : ∀ j, ∑ i, v i * (f j ^ (i : ℕ)) = 0) : v = 0 :=
+by { refine eq_zero_of_forall_index_sum_pow_mul_eq_zero hf _, simp_rw mul_comm, exact hfv }
```
Does this work?
```suggestion
by { apply eq_zero_of_forall_index_sum_pow_mul_eq_zero hf, simp_rw mul_comm, exact hfv }
```

vihdzp reviewed (commented):
A few general remarks.

linesthatinterlace committed:
Changes after review

linesthatinterlace committed:
Fix line length

linesthatinterlace committed:
Fix lint and style

Vierkantor commented on src/linear_algebra/lagrange.lean:
```diff
@@ -1,224 +1,566 @@
+@[simp] theorem basis_pair_left (hij : i ≠ j) :
+  lagrange.basis {i, j} v i = basis_divisor (v i) (v j) :=
+by { simp only [lagrange.basis, hij, erase_insert_eq_erase, erase_eq_of_not_mem,
+                mem_singleton, not_false_iff, prod_singleton] }
```
```suggestion
by simp only [lagrange.basis, hij, erase_insert_eq_erase, erase_eq_of_not_mem,
              mem_singleton, not_false_iff, prod_singleton]
```

Vierkantor commented on src/linear_algebra/lagrange.lean:
```diff
@@ -1,224 +1,566 @@
+       ← mul_assoc (x - v i)⁻¹, inv_mul_cancel (sub_ne_zero_of_ne hxi), one_mul]
+
+lemma interpolate_eq_nodal_weight_mul_nodal_div_X_sub_C :
+interpolate s v r = ∑ i in s, C (nodal_weight s v i) * (nodal s v / (X - C (v i))) * C (r i) :=
```
```suggestion
  interpolate s v r = ∑ i in s, C (nodal_weight s v i) * (nodal s v / (X - C (v i))) * C (r i) :=
```

Vierkantor reviewed (approved):
Everything looks good with just some minor remarks, thanks!

bors d+

linesthatinterlace committed:
Apply suggestions from code review

Co-authored-by: Anne Baanen <Vierkantor@users.noreply.github.com>
